examples/virtual-dom-wip/platform/Html/Internal/Client.roc

```diff
@@ -0,0 +1,538 @@
+
+    {
+        rendered: renderedChild,
+        patches: childPatches,
+        ids: List.append ids id,
+    }
+
+# insert a node into the nodes list, assigning it a NodeId
+insertNode : RenderedTree state, RenderedNode -> { rendered : RenderedTree state, id : NodeId }
+insertNode = \rendered, node ->
+    when List.last rendered.deletedNodeCache is
+        Ok id ->
+            newRendered =
+                { rendered &
+                    nodes: List.set rendered.nodes id (Ok node),
+                    deletedNodeCache: List.dropLast rendered.deletedNodeCache,
+                }
+
+            { rendered: newRendered, id }
+
+        Err _ ->
+            newRendered =
+                { rendered &
+                    nodes: List.append rendered.nodes (Ok node),
+                }
+
+            { rendered: newRendered, id: List.len rendered.nodes }
+
+# Predict what NodeId will be assigned next, without actually assigning it
+nextNodeId : RenderedTree state -> NodeId
+nextNodeId = \rendered ->
+    when List.last rendered.deletedNodeCache is
+        Ok id -> id
+        Err _ -> List.len rendered.nodes
+
+# -------------------------------
+#   TESTS
+# -------------------------------
+# indexNodes
+expect
+    html : Html {}
+    html =
+        Element "a" 43 [HtmlAttr "href" "https://www.roc-lang.org/"] [Text "Roc"]
+
+    actual : { nodes : List RenderedNode, siblingIds : List Nat }
+    actual =
+        indexNodes { nodes: [], siblingIds: [] } html
+
+    expected : { nodes : List RenderedNode, siblingIds : List Nat }
+    expected = {
+        nodes: [
+            RenderedText "Roc",
+            RenderedElement "a" [RenderedHtmlAttr "href" "https://www.roc-lang.org/"] [0],
+        ],
+        siblingIds: [1],
+    }
+
+    actual == expected
+
+# initClientAppHelp
+expect
+    State : { answer : Nat }
+
+    init = \result ->
+        when result is
+            Ok state -> state
+            Err _ -> { answer: 0 }
+
+    onClickHandler : Handler State
+    onClickHandler =
+        Normal \state, _ -> Action.update { answer: state.answer + 1 }
+
+    render : State -> Html State
+    render = \state ->
+        num = Num.toStr state.answer
+
+        onClickAttr : Attribute State
+        onClickAttr =
+            EventListener "click" [] onClickHandler
+
+        Element "body" 0 [] [
+            Element "h1" 0 [] [Text "The app"],
+            Element "div" 0 [onClickAttr] [Text "The answer is \(num)"],
+        ]
+
+    app : App State State
+    app = {
+        init,
+        render,
+        wasmUrl: "assets/test.wasm",
+    }
+
+    initJson : List U8
+    initJson =
+        # { answer: 42 } |> Encode.toBytes Json.toUtf8 # panics at mono/src/ir.rs:5739:56
+        "{ answer: 42 }" |> Str.toUtf8
+
+    # COMPILER BUG? 'no lambdaset found'
+    actual : { state : State, rendered : RenderedTree State, patches : List Patch }
+    actual =
+        initClientAppHelp initJson app
```

Comment:
This passes `roc check` but fails `roc test`.
I get the same error when I do `roc build examples/virtual-dom-wip/example-client.roc`

Reply:
This test isn't actually running in CI yet so we could merge as is if there's no easy way to fix it.

Reply:
@ayazhafiz do you think this is related to any known bug?

It's throwing here
https://github.com/roc-lang/roc/blob/f2d6def117f4d9e25a95b016cc88b39f60a5f130/crates/compiler/mono/src/layout.rs#L1411

Reply:
Ah with a debug build it changes to
```
function symbol `Html.Internal.Client.createChildNode` not in set LambdaSet { set: [], representation: Interned(1, PhantomData) }
```

Reply:
Hmm. I'm not aware of issues similar to this yet. Let's get a minimal example as in #4806 first, and also check if this is resolved by #4819 after that lands.